### ISSUE TITLE
boundary calculation too slow for weather routing

### ISSUE THREAD

**Issue by seandepagnier** [jongough/ocpn_draw_pi#257]
Jon,

I am not sure how exactly you want to resolve this, but the current method using json messages is far too slow for practical use.

First of all, the boundary test routine should not take a single point, but a segment, because it is currently possible for a weather route to cut across the boundary (cut the corner) as the segment end points are both outside the boundary.

Secondly, I recommend using a single message to pass a function pointer for boundary segment tests.  The actual boundary segment test routine needs to hash boundaries by some kind of lat/lon grid or possibly binary tree to greatly speed the search.

I am considering developing some sort of datastructure to contain lat/lon bounding box objects for log(n)  test time to consider which objects are in a certain viewport (or could apply to a segment)

What do you want to do about it?  It would be great to have boundary support in weather routing.

**Comment by jongough** [comment]
Sean,
   In the current beta (release_1.1 branch, although I seem to have committed changes to master by mistake, so am trying to fix that) there is a new message, FindClosestBoundaryLineCrossing, which takes a line segment and sees if it crosses a boundary. I have implemented this into the watchdog plugin which will use the new code if the OD version is 1.1.1 or greater. I did not create a pull request for the weather routing as I thought you were changing it and did not want any more changes from me.

I am not really sure what you are suggesting, but it sounds like using external modules, i.e. tightly coupling the plugins together. I thought the reason for JSON messaging was to avoid this. I do believe that the current messaging needs to be more streamlined, made asynchronous with targeted messages rather than the current spray method. I have raised the messaging and I thought it was going to be investigated for OCPN 5.x.

Boundary support is currently in weather routing, just it uses points rather than lines. I can provide a patch that will use the line function if you want.

**Comment by seandepagnier** [comment]
Jon,

I am finished with my weather routing changes.  I didn't want changes because I was on passage in the atlantic and without internet for weeks there would be conflicts.  Now I have internet access so it's not a problem.

I agree that tightly coupling the plugins is not so nice, but the speed of json parsing is simply not acceptable.  Try it for yourself and give weather routing a boundary to avoid and compare it to without the boundaries.  I don't see much problem in simply passing a function pointer in a json message once, then calling this function repeatedly.

As far as message subscriptions or mailboxes or whatever, perhaps it would be a nice thing and could give a performance improvement, sure, but it's still not going to be sufficient for this unless we also have binary messages or something.

**Comment by jongough** [comment]
I originally followed the OCPN method and exposed an API that could be used without the need for JSON messaging, but I removed it as it seemed that JSON was the preferred method of communicating. 

Would it be better if I introduced the API again and then you would use a new header file declaring the functions that are externally exposed, rather like OCPN? This would allow compile time 'linking' and then just use the JSON message to find out if OD was available and of the requisite version, if all is good then call the needed function.

**Comment by seandepagnier** [comment]
sure we could do that way...  I think I was passing function pointers
over json as well but maybe you are right that the dynamic linker can
take care of it.

**Comment by jongough** [comment]
I am completely new at this game. I think we could do this if we could gain access to the pluginmanager list of plugin addresses, but then I would need to provide the API in the ocpn_plugin.h file. I am hoping that there is a way to use functions from another DLL without having to load the DLL again AND be able to access its functions. But I am struggling a bit here, so any idea how to do it.

I don't particularly like the idea of passing function addresses in pointers when they are not in the same executable.

**Comment by seandepagnier** [comment]
I am confused by your last statement.

I pass some function pointers from climatology to weather routing as
you can see in weather_routing_pi.cpp:260

We could simply pass the boundary test routine to weather routing via
a message storing the function pointer as a string in the json
message.  Once weather routing has the function pointer it can call it
thousands of times without all the overhead of parsing json.

**Comment by rgleason** [comment]
```
      wxString sptr = v[_T("ClimatologyDataPtr")].AsString();
        sscanf(sptr.To8BitData().data(), "%p", &RouteMap::ClimatologyData);

        sptr = v[_T("ClimatologyWindAtlasDataPtr")].AsString();
        sscanf(sptr.To8BitData().data(), "%p", &RouteMap::ClimatologyWindAtlasData);

        sptr = v[_T("ClimatologyCycloneTrackCrossingsPtr")].AsString();
        sscanf(sptr.To8BitData().data(), "%p", &RouteMap::ClimatologyCycloneTrackCrossings);

        if(m_pWeather_Routing) {
            m_pWeather_Routing->m_ConfigurationDialog.m_cClimatologyType->Enable
                (RouteMap::ClimatologyData!=NULL);
            m_pWeather_Routing->m_ConfigurationDialog.m_pCyclones->Enable
                (RouteMap::ClimatologyCycloneTrackCrossings!=NULL);
```

**Comment by seandepagnier** [comment]
I suppose it could also be done via dlopen, but I still think it's safer to pass the pointer so we can do version checking between the plugins.

Anyway, I will try to make a PR to ocpn_draw_pi

**Comment by jongough** [comment]
I have created the JSON message, GetAPIAddresses, which will return all the API's being made available this way. Currently only "OD_FindPathByGUID", but I will extend that as soon as I get it all working.

**Comment by jongough** [comment]
I have provided a patch to watchdog that uses this API to get the GUID. It will still work if the new API is not available, but will use the API if it is. This is more a proof of concept to show it working. Which API/JSON message would you like me to implement first?

**Comment by seandepagnier** [comment]
Jon,

Great work.  I need ne identical to the gshhs crossing routine, so
using segments not an end point.   The boundaries themselves should
have lat/lon bounding boxes to speed up tests considerably.

They could also be placed into a quad tree to support thousands of
boundaries... but don't worry about this bit just yet.

Leave the weather routing change to me, I will take care of it.

**Comment by jongough** [comment]
I have just put up OD 1.1.17 which contains new API calls, FindPointInAnyBoundary and FindClosestBoundaryLineCrossing. I have also created a pull request for the watchdog plugin with these API calls being used. I have not accepted these as I think you, Sean, should check it out and see if you are happy with the implementation.

The bounding box could be implemented, but would it speed up testing much? I do not envisage, at the moment hundreds of boundaries, may be 1-20, but if users start to use large numbers then it may be worth it.

Jon

**Comment by seandepagnier** [comment]
It would speed things up considerably to use the bounding box because
it is only 1-4 comparisons to determine the boundary is not going to
touch.   If you had hundreds of boundaries then they should also be in
a quad tree.

I am looking for your work in ocpn_draw_pi and not finding it when I
pulled.  I don't find it in watchdog neither.

Sean

**Comment by jongough** [comment]
These changes are currently going in to the release_1.1 branch of OD. The changes for WD are in my clone, I have just created the pull request, forgot to do it last night, sorry. I am not merging these with 'master' as I wanted you to approve or reject them. I think they are getting a little scrappy as now checking for version of OD for two different changes is making it messy. I am hoping that when OD 1.2 goes out the door (1.1 being the beta for 1.2) I will be able to clean some of it up.

The bounding box stuff will only be of benefit if you are checking for a particular point being in a boundary, with line crossing the boundary itself is the bounding box. This has reduced the number of checks that are needed considerably. I will have a think and see if I can work out a way to make this work, or do you have an algorithm for lines crossing bounding boxes that I can use. I am already doing vector checking for lines crossing boundary lines.

**Comment by jongough** [comment]
Do you have any 'bounding box' logic I could have a look at to see if it will help with identifying boundaries to be checked? The issue I see is that using a vector rather than a point both end points may be outside the bounding box but the line joining the points may cross the box. Does the bounding box logic cover this?

**Comment by seandepagnier** [comment]
It can cover this.  I did a lot of work improving LLBBox class but
it's still a pull request for about 2 months now.  The old code is
still in master which works it's just less efficient.

 Assuming latlon boxes a.re either from -180 to 180, 360 to 0, or 0 to
360 in all cases You want to minimize the number of comparisons, so
for a point:

Test if point latitude is too low or high and return false

if the box max lon is > 180 and the point lon is < 0, add 360 to point lon
else if the min lon is < -180 and the point lon is > 0 subtract 360 to point lon

Test if point longitude is too high or low and return false

Now return what the normal boundary test gives.


Basically, if the point is on the wrong side of any of the 4 lines
then it must be outside.  The same holds true for a vector, but in
this case both points must be on the wrong side of the _same_ line.
If one is too much lat, and the other too much lon a full test still
must take place.


For a vector you need double the comparisons with the same logic

Test if point 1 and point 2 have latitude that is too high, if so, return false
Test if point 1 and point 2 have latitude that is too low, if so, return false

normalize longitude...

lon tests,

vector boundary test


I hope it helps let me know if you have any questions or ideas to
speed things up.

**Comment by rgleason** [comment]
Just out of curiosity is sean's box test now implemented?
Sounds like it is useful for many things, testing Land, testing GZ, etc.
Is there some way to package it as a function?

**Comment by jongough** [comment]
I have not implemented this yet. It may be worthwhile but I am not sure yet. I will not be implementing this until the 1.4 release (1.3 alpha/beta is likely to be Nov).
